Fix Euler's method (non-)initialization in Burgers

wthrowe opened the pull request:
The other systems already worked.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Fix Euler's method (non-)initialization in Burgers

The other systems already worked.

nilsvu commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -158,7 +158,8 @@ struct EvolutionMetavars {
               Parallel::PhaseActions<
                   Phase, Phase::RegisterWithObserver,
-                  tmpl::list<observers::Actions::RegisterWithObservers<
+                  tmpl::list<Actions::AdvanceTime,
```
Minor suggestion: Can you do this at the end of the inititalization phase instead of in the registering phase? Makes more sense semantically

nilsdeppe replied:
That's not the same place in the algorithm... Registration happens after time stepper initialization.

nilsvu replied:
Right, so this would be better placed at the end of `Phase::InitializeTimeStepperHistory` I think. Or is it placed in the register phase for a reason?

wthrowe replied:
I think it would make more sense in the time-stepper initialization, but moving it there is non-trivial.  The modifications to the TimeId would have to occur in the middle of the Cleanup action, so that action would have to be split.  This is not particularly difficult, but it is more than I want to get into in this PR.

nilsvu replied:
Is the issue that the time-stepper initialization phase is a loop and you want to invoke AdvanceTime only when it’s finished? In that case, how about you add a new phase for this like `Phase::PrepareEvolution`?

kidder replied:
we want to minimize the number of phases as much as possible

nilsvu replied:
Sure. I don’t think a phase like this would have any significant impact on runtime though. Anyway, perhaps a comment in the code could help readers understand the reasoning you described here. Please consider this suggestion non-blocking for merging this PR.

nilsdeppe commented:
My concern with more phases isn't just performance but complexity in
reasoning about the code. Using a phase for this is like using a sledge
hammer to kill a fly:it's technically possible but certainly not the right
tool. I view using a separate phase here as going from understandable code
to a big hack to deal with a minor "issue".

 I also don't think a comment in burgers would be useful. This affects all
evolution systems and should be changed in a separate PR, if at all.

I'm with Will, this should be merged as is. I think too much time has
already been invested into such a minor thing.

Cheers,

Nils

On Tue, Jul 16, 2019, 22:43 Nils Leif Fischer <notifications@github.com>
wrote:

> *@nilsleiffischer* commented on this pull request.
> ------------------------------
>
> In src/Evolution/Executables/Burgers/EvolveBurgers.hpp
> <https://github.com/sxs-collaboration/spectre/pull/1604#discussion_r304199698>
> :
>
> > @@ -158,7 +158,8 @@ struct EvolutionMetavars {
>
>
>
>                Parallel::PhaseActions<
>
>                    Phase, Phase::RegisterWithObserver,
>
> -                  tmpl::list<observers::Actions::RegisterWithObservers<
>
> +                  tmpl::list<Actions::AdvanceTime,
>
>
> Sure. I don’t think a phase like this would have any significant impact on
> runtime though. Anyway, perhaps a comment in the code could help readers
> understand the reasoning you described here. Please consider this
> suggestion non-blocking for merging this PR.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sxs-collaboration/spectre/pull/1604?email_source=notifications&email_token=ADHHPFTS7KMQMMX7B2WMZY3P72BN7A5CNFSM4ID2F43KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6U22BI#discussion_r304199698>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADHHPFTFMTG4N5ENBSI437TP72BN7ANCNFSM4ID2F43A>
> .
>

wthrowe commented:
How about we hold off merging this and I'll fix the self-start code and then you all can stop arguing about it?

nilsvu commented:
As I wrote earlier, I’m fine with merging this as is and my comments were non-blocking and meant to understand why this is done the way it is. I think this is justified when there is an `AdvanceTime` action in a `RegisterWithObserver` phase :)

wthrowe committed:
Remove unnecessary DataBox content checks

The self-start actions should only be used under very controlled
circumstances where we know how the DataBox type will change
throughout the phase.

wthrowe committed:
Fix comments not matching the PDALs implementation

wthrowe committed:
Be more careful testing the self-start final state

Remove the assumption that Cleanup is the end of the procedure.  Also
add a check on the TimeId.

wthrowe committed:
Initialize time evolution in EvolveScalarWave

We really need some way to test that these executables do something
beyond "not crashing".

wthrowe committed:
Set initial time at the end of self-starting

wthrowe commented:
Done.  I also fixed EvolveScalarWave, which appears to have been broken without anyone noticing since PDALs were merged.

nilsvu reviewed (approved):
LGTM 👍

nilsvu commented:
And yes, we definitely need some kind of regression tests for our executables..
